pkg/resources/secret.go

```diff
@@ -0,0 +1,192 @@
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Secret struct {
+	Wireguard *v1alpha1.Wireguard
+	Logger    logr.Logger
+	Client    client.Client
+	Scheme    *runtime.Scheme
+}
+
+func (s Secret) Exists(ctx context.Context) (bool, error) {
+
+	secret := &corev1.Secret{}
+	err := s.Client.Get(ctx, types.NamespacedName{Name: s.Name(), Namespace: s.Wireguard.Namespace}, secret)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (s Secret) Converged(ctx context.Context) (bool, error) {
+	return true, nil
+}
+
+func (s Secret) NeedsUpdate(ctx context.Context) (bool, error) {
+
+	secret := &corev1.Secret{}
+	err := s.Client.Get(ctx, types.NamespacedName{Name: s.Name(), Namespace: s.Wireguard.Namespace}, secret)
+
+	if err != nil {
+		return true, err
+	}
+
+	expectedSecret, err := s.getSecreteData(ctx)
+
+	if err != nil {
+		return true, err
+	}
+
+	if !bytes.Equal(expectedSecret.Data["state.json"], secret.Data["state.json"]) {
+		return true, nil
+
+	}
+
+	return false, nil
+}
+
+func (s Secret) Update(ctx context.Context) error {
+	sec, err := s.getSecreteData(ctx)
+
+	if err != nil {
+		return err
+	}
+	if err := s.Client.Update(ctx, sec); err != nil {
+		s.Logger.Error(err, fmt.Sprintf("unable to update secret %v and %s", err, s.Name()))
+		return err
+
+	}
+
+	return nil
+}
+
+func (s Secret) Create(ctx context.Context) error {
+	sec, err := s.getSecreteData(ctx)
+
+	if err != nil {
+		return err
+	}
+	if err := s.Client.Create(ctx, sec); err != nil {
+		return err
+
+	}
+	return nil
+}
+func (s Secret) Type() string {
+	return "secret"
+}
+
+func (s Secret) Name() string {
+	return fmt.Sprintf("%s-%s", s.Wireguard.Name, s.Wireguard.Status.UniqueIdentifier)
+}
+
+func (s Secret) GetResourceVersion(ctx context.Context) string {
+	sec, err := s.getExistingSecret(ctx)
+	if err != nil {
+		return "1"
+	}
+	return sec.ResourceVersion
+}
+
+func (s Secret) GetPublicKey(ctx context.Context) (string, error) {
+	sec, err := s.getSecreteData(ctx)
+	if err != nil {
+		return "", err
+	}
+
+	return string(sec.Data["publicKey"]), nil
+}
+func (s Secret) getSecreteData(ctx context.Context) (*corev1.Secret, error) {
+
+	data := map[string][]byte{}
+
+	peers, err := s.getPeersInfo(ctx)
+	if err != nil {
+		return &corev1.Secret{}, err
+	}
+
+	sec, err := s.getExistingSecret(ctx)
+	publicKey := ""
+	privateKey := ""
+
+	if errors.IsNotFound(err) {
+
+		key, err := wgtypes.GeneratePrivateKey()
+		if err != nil {
+			return &corev1.Secret{}, err
+		}
+
+		privateKey = key.String()
+		publicKey = key.PublicKey().String()
+
+	} else if err == nil {
+		privateKey = string(sec.Data["privateKey"])
+		publicKey = string(sec.Data["publicKey"])
```

Comment:
This should be removed, as everything preceding the `if err != nil` line matches this condition